integration/tests/services/test_ws_subscriber.py

```diff
@@ -0,0 +1,281 @@
+        instruction_tx = event_caller.functions.callEvent2(arg1, arg2).build_transaction(
+            tx
+        )
+        self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+        tx = self.make_tx_object(self.sender_account.address)
+        instruction_tx = event_caller.functions.callEvent2(arg2, arg3).build_transaction(
+            tx
+        )
+        self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+        tx = self.make_tx_object(self.sender_account.address)
+        instruction_tx = event_caller.functions.callEvent3(
+            arg1, arg2, arg3
+        ).build_transaction(tx)
+        self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+    @staticmethod
+    def assert_all_messages(messages: list, topics: list):
+        is_event_topic_in_list = False
+        is_arg_topic_in_list = False
+        for m in messages:
+            log = m.params.result
+            if topics[0]:
+                for topic in topics[0]:
+                    if topic in log.topics:
+                        is_event_topic_in_list = True
+            else:
+                is_event_topic_in_list = True
+            if len(topics) == 2:
+                for topic in topics[1]:
+                    if topic in log.topics:
+                        is_arg_topic_in_list = True
+            else:
+                is_arg_topic_in_list = True
+
+        assert is_event_topic_in_list, f"Filter by {topics} works incorrect. Response: {m}"
+        assert is_arg_topic_in_list, f"Filter by {topics} works incorrect. Response: {m}"
+
+    async def test_subscribe_to_newheads(self):
+        async with websockets.connect(TEST_STAND) as ws:
+            data = Subscribe(["newHeads"])
+            await ws.send(json.dumps(data.__dict__))
+            r = await ws.recv()
+            response = json.loads(r, object_hook=lambda d: SimpleNamespace(**d))
+            subscription = response.result
+            assert len(subscription) > 0
+
+            self.send_neon(self.sender_account, self.recipient_account, 0.1)
+
+            r = await ws.recv()
+            response = json.loads(r, object_hook=lambda d: SimpleNamespace(**d))
+            assert hasattr_recursive(response, "params.subscription")
+            assert hasattr(response.params, "result")
+            assert response.params.subscription == subscription
+            assert_fields_are_hex(response.params.result,
+                                  ["extraData", "gasLimit", "gasUsed", "logsBloom",
+                                   "nonce", "parentHash", "sha3Uncles",
+                                   "stateRoot", "timestamp", "transactionsRoot"])
+
+            data = Unsubscribe(subscription)
+            await ws.send(json.dumps(data.__dict__))
+            messages = await ws_receive_all_messages(ws)
+            response = messages[len(messages) - 1]
+            assert response.result == 'true'
```

Comment:
you don't check transactions inside newHeads and block number don't compare with block number from transaction receipt. But in real situation for small time you can get several blocks and you should check it in cycle and found head with your transaction and your block number
https://docs.alchemy.com/reference/newheads

Reply:
Added check by `blockHash`.